crypto/s2n_prf_libcrypto.c

```diff
@@ -0,0 +1,192 @@
+            RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&seed_b_stuffer, seed_b->data, seed_b->size));
+            RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&seed_b_stuffer, seed_c->data, seed_c->size));
+        }
+
+        seed_b_len = s2n_stuffer_data_available(&seed_b_stuffer);
+        seed_b_data = s2n_stuffer_raw_read(&seed_b_stuffer, seed_b_len);
+        RESULT_ENSURE_REF(seed_b_data);
+    }
+
+    RESULT_GUARD_OSSL(CRYPTO_tls1_prf(digest,
+                              out->data, out->size,
+                              secret->data, secret->size,
+                              (const char *) label->data, label->size,
+                              seed_a->data, seed_a->size,
+                              seed_b_data, seed_b_len),
+            S2N_ERR_PRF_DERIVE);
+
+    return S2N_RESULT_OK;
+}
+
+#elif S2N_OPENSSL_VERSION_AT_LEAST(3, 0, 0)
+
+    #include <openssl/core_names.h>
+    #include <openssl/kdf.h>
+
+    #define OSSL_PARAM_BLOB(id, blob) \
+        OSSL_PARAM_octet_string(id, blob->data, blob->size)
+    #define OSSL_PARAM_STR(id, cstr) \
+        OSSL_PARAM_utf8_string(id, cstr, 0)
+
+DEFINE_POINTER_CLEANUP_FUNC(EVP_KDF_CTX *, EVP_KDF_CTX_free);
+DEFINE_POINTER_CLEANUP_FUNC(EVP_KDF *, EVP_KDF_free);
+
+S2N_RESULT s2n_prf_libcrypto(struct s2n_connection *conn,
+        struct s2n_blob *secret, struct s2n_blob *label,
+        struct s2n_blob *seed_a, struct s2n_blob *seed_b, struct s2n_blob *seed_c,
+        struct s2n_blob *out)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(secret);
+    RESULT_ENSURE_REF(label);
+    RESULT_ENSURE_REF(seed_a);
+    RESULT_ENSURE_REF(out);
+
+    struct s2n_blob empty_seed = { 0 };
+    if (!seed_b) {
+        seed_b = &empty_seed;
+    }
+    if (!seed_c) {
+        seed_c = &empty_seed;
+    }
+
+    /* Openssl limits the size of the seed to 1024 bytes, including the label.
+     * This would be an issue for TLS1.2 PQ, which uses full keyshares as seeds.
+     * However, s2n-tls doesn't support PQ with Openssl, so this limitation will
+     * never affect customers.
+     *
+     * As of this commit, EVP_KDF_derive will fail silently (without logging any
+     * error) if the seed is too large. This check adds visibility.
+     */
+    uint64_t seed_total_size = label->size + seed_a->size + seed_b->size + seed_c->size;
+    RESULT_ENSURE(seed_total_size <= 1024, S2N_ERR_PRF_INVALID_SEED);
+
+    const char *digest_name = "MD5-SHA1";
+    const char *fetch_properties = "-fips";
+
+    if (conn->actual_protocol_version == S2N_TLS12) {
+        fetch_properties = NULL;
+
+        RESULT_ENSURE_REF(conn->secure);
+        RESULT_ENSURE_REF(conn->secure->cipher_suite);
+        s2n_hmac_algorithm prf_alg = conn->secure->cipher_suite->prf_alg;
+
+        const EVP_MD *digest = NULL;
+        RESULT_GUARD(s2n_hmac_md_from_alg(prf_alg, &digest));
+        RESULT_ENSURE(digest, S2N_ERR_PRF_INVALID_ALGORITHM);
```

Comment:
nit - this error is kind of misleading, since `s2n_hmac_md_from_alg()` will raise the error if `prf_alg` is invalid.
```suggestion
        RESULT_ENSURE_REF(digest);
```